java.lang.NullPointerException: Cannot invoke "String.length()" because "plainText" is null

### Is there an existing issue for this?

- [x] I have searched the existing issues and didn't find mine.

### Steps to reproduce

The software has already worked on my pc, I tried to restart it this morning with a 'maestro test ...' and I got this error. I tried to restart nothing happened.

### Actual results

java.lang.NullPointerException: Cannot invoke "String.length()" because "plainText" is null
        at picocli.CommandLine$Help$ColorScheme.apply(CommandLine.java:17173)
        at picocli.CommandLine$Help$ColorScheme.errorText(CommandLine.java:17049)
        at maestro.cli.AppKt.main$lambda$3(App.kt:128)
        at picocli.CommandLine.execute(CommandLine.java:2088)
        at maestro.cli.AppKt.main(App.kt:143)

### Expected results

this commands already work

### About app

I use the release latest version of maestro for windows
Framework used to build the app:
React Native


### About environment

java version "22.0.1" 2024-04-16
windows 11 Version 10.0.22631 Build 22631
x86_64

### Logs

maestro logfile is blank


### Maestro version

1.39.9

### How did you install Maestro?

other (please specify in the text area below)

### Anything else?

zip software for windows

Reproduced on windows git bash. Somewhat flaky behavior when using --verbose. 

```as@asko-22 MINGW64 /c/git/maestro-testing/samples
$ which maestro
/c/maestro/bin/maestro

as@asko-22 MINGW64 /c/git/maestro-testing/samples
$ maestro --version
1.39.13

as@asko-22 MINGW64 /c/git/maestro-testing/samples
$ adb install sample.apk
Performing Streamed Install
Success

as@asko-22 MINGW64 /c/git/maestro-testing/samples
$ maestro test android-flow.yaml
java.lang.NullPointerException: Cannot invoke "String.length()" because "plainText" is null
        at picocli.CommandLine$Help$ColorScheme.apply(CommandLine.java:17173)
        at picocli.CommandLine$Help$ColorScheme.errorText(CommandLine.java:17049)
        at maestro.cli.AppKt.main$lambda$3(App.kt:129)
        at picocli.CommandLine.execute(CommandLine.java:2088)
        at maestro.cli.AppKt.main(App.kt:144)```

Regular windows command prompt
```
C:\git\maestro-testing\samples>c:\maestro\bin\maestro --verbose test android-flow.yaml
[ INFO] ---- System Info ----
[ INFO] Maestro Version: 1.39.13
[ INFO] CI: Undefined
[ INFO] OS Name: Windows 11
[ INFO] OS Version: 10.0
[ INFO] Architecture: amd64
[ INFO] Java Version: 17
[ INFO] Xcode Version: null
[ INFO] Flutter Version: Undefined
[ INFO] Flutter Channel: Undefined
[ INFO] ---------------------
[DEBUG] Using SLF4J as the default logging framework
[ INFO] [shard 1] Selected device emulator-5554 using port 7001
java.lang.NullPointerException: Cannot invoke "String.length()" because "plainText" is null
        at picocli.CommandLine$Help$ColorScheme.apply(CommandLine.java:17173)
        at picocli.CommandLine$Help$ColorScheme.errorText(CommandLine.java:17049)
        at maestro.cli.AppKt.main$lambda$3(App.kt:129)
        at picocli.CommandLine.execute(CommandLine.java:2088)
        at maestro.cli.AppKt.main(App.kt:144)

C:\git\maestro-testing\samples>c:\maestro\bin\maestro --verbose test android-flow.yaml
[ INFO] ---- System Info ----
[ INFO] Maestro Version: 1.39.13
[ INFO] CI: Undefined
[ INFO] OS Name: Windows 11
[ INFO] OS Version: 10.0
[ INFO] Architecture: amd64
[ INFO] Java Version: 17
[ INFO] Xcode Version: null
[ INFO] Flutter Version: Undefined
[ INFO] Flutter Channel: Undefined
[ INFO] ---------------------
[DEBUG] Using SLF4J as the default logging framework
[ INFO] [shard 1] Selected device emulator-5554 using port 7001

Running on emulator-5554

 ║
 ║  > Flow: android-flow
 ║
 ║        Launch app "org.wikipedia"
 ║
[ INFO] Getting device info
[ERROR] Failed to run flow
io.grpc.StatusRuntimeException: UNAVAILABLE: io exception
        at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:275)
        at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:256)
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:169)
        at maestro_android.MaestroDriverGrpc$MaestroDriverBlockingStub.deviceInfo(MaestroDriverGrpc.java:634)
        at maestro.drivers.AndroidDriver$deviceInfo$1.invoke(AndroidDriver.kt:185)
        at maestro.drivers.AndroidDriver$deviceInfo$1.invoke(AndroidDriver.kt:184)
        at maestro.drivers.AndroidDriver.runDeviceCall(AndroidDriver.kt:1147)
        at maestro.drivers.AndroidDriver.deviceInfo(AndroidDriver.kt:184)
        at maestro.Maestro$cachedDeviceInfo$2.invoke(Maestro.kt:51)
        at maestro.Maestro$cachedDeviceInfo$2.invoke(Maestro.kt:49)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
        at maestro.Maestro.getCachedDeviceInfo(Maestro.kt:49)
        at maestro.orchestra.Orchestra.initJsEngine(Orchestra.kt:240)
        at maestro.orchestra.Orchestra.runFlow(Orchestra.kt:118)
io.grpc.StatusRuntimeException: UNAVAILABLE: io exceptionmands(MaestroCommandRunner.kt:187)
        at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:275)4)
        at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:256)tRunner.kt:55)
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:169)
        at maestro_android.MaestroDriverGrpc$MaestroDriverBlockingStub.deviceInfo(MaestroDriverGrpc.java:634)
        at maestro.drivers.AndroidDriver$deviceInfo$1.invoke(AndroidDriver.kt:185)
        at maestro.drivers.AndroidDriver$deviceInfo$1.invoke(AndroidDriver.kt:184)
        at maestro.drivers.AndroidDriver.runDeviceCall(AndroidDriver.kt:1147).kt:353)
        at maestro.drivers.AndroidDriver.deviceInfo(AndroidDriver.kt:184)mand.kt:323)
        at maestro.Maestro$cachedDeviceInfo$2.invoke(Maestro.kt:51)stroSessionManager.kt:105)
        at maestro.Maestro$cachedDeviceInfo$2.invoke(Maestro.kt:49)ault(MaestroSessionManager.kt:53)
        at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)ommand.kt:323)
        at maestro.Maestro.getCachedDeviceInfo(Maestro.kt:49)te(TestCommand.kt:65)
        at maestro.orchestra.Orchestra.initJsEngine(Orchestra.kt:240)$1.invokeSuspend(TestCommand.kt:290)
        at maestro.orchestra.Orchestra.runFlow(Orchestra.kt:118)esumeWith(ContinuationImpl.kt:33)
        at maestro.cli.runner.MaestroCommandRunner.runCommands(MaestroCommandRunner.kt:187)
        at maestro.cli.runner.TestRunner$runSingle$result$1.invoke(TestRunner.kt:64)r.kt:111)
        at maestro.cli.runner.TestRunner$runSingle$result$1.invoke(TestRunner.kt:55)
        at maestro.cli.runner.TestRunner.runCatching(TestRunner.kt:165)oroutineScheduler.kt:585)
        at maestro.cli.runner.TestRunner.runSingle(TestRunner.kt:55)xecuteTask(CoroutineScheduler.kt:802)
        at maestro.cli.command.TestCommand.runSingleFlow(TestCommand.kt:382)(CoroutineScheduler.kt:706)
        at maestro.cli.command.TestCommand.access$runSingleFlow(TestCommand.kt:65)heduler.kt:693)
Caused bat maestro.cli.command.TestCommand$runShardSuite$2.invoke(TestCommand.kt:353)used: no further information: localhost/[0:0:0:0:0:0:0:1]:7001
Caused bat maestro.cli.command.TestCommand$runShardSuite$2.invoke(TestCommand.kt:323)
        at maestro.cli.session.MaestroSessionManager.newSession(MaestroSessionManager.kt:105)
        at maestro.cli.session.MaestroSessionManager.newSession$default(MaestroSessionManager.kt:53)
        at maestro.cli.command.TestCommand.runShardSuite(TestCommand.kt:323)lImpl.java:946)
        at maestro.cli.command.TestCommand.access$runShardSuite(TestCommand.kt:65)hannel.java:337)
        at maestro.cli.command.TestCommand$handleSessions$1$results$1$1.invokeSuspend(TestCommand.kt:290)a:334)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)ed(NioEventLoop.java:724)
        at kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:111)
        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:99):562)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:585)va:997)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:802)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:706)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:693)
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: no further information: localhost/[0:0:0:0:0:0:0:1]:7001
Caused by: java.net.ConnectException: Connection refused: no further information
        at java.base/sun.nio.ch.Net.pollConnect(Native Method)
        at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
        at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:946)
        at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:337)
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:776)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)


==== Debug output (logs & screenshots) ====

C:\Users\as\.maestro\tests\2025-03-04_165446
[ERROR] Failed to record heartbeat
java.lang.InterruptedException: sleep interrupted
        at java.base/java.lang.Thread.sleep(Native Method)
        at maestro.cli.session.MaestroSessionManager.newSession$lambda$0(MaestroSessionManager.kt:75)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

--- a/maestro-cli/src/main/java/maestro/cli/App.kt
+++ b/maestro-cli/src/main/java/maestro/cli/App.kt
@@ -126,7 +126,7 @@ fun main(args: Array<String>) {
                 .build()
 
             cmd.err.println(
-                cmd.colorScheme.errorText(ex.message)
+                cmd.colorScheme.errorText(ex.message.orEmpty())
             )
 
             if (ex !is CliError && ex !is MaestroException.UnsupportedJavaVersion) {

--- a/maestro-cli/src/main/java/maestro/cli/util/DeviceConfig.kt
+++ b/maestro-cli/src/main/java/maestro/cli/util/DeviceConfig.kt
@@ -28,19 +28,19 @@ data class DeviceConfigAndroid(
         val versions = listOf(34, 33, 31, 30, 29, 28)
         val defaultVersion = 30
 
-        fun createConfig(version: Int, device: AvdDevice, architecture: MACOS_ARCHITECTURE): DeviceConfigAndroid {
+        fun createConfig(version: Int, device: AvdDevice, architecture: CPU_ARCHITECTURE): DeviceConfigAndroid {
             val name = "Maestro_${device.name.replace(" ", "_")}_API_${version}"
             val tag = "google_apis"
             val systemImage = when (architecture) {
-                MACOS_ARCHITECTURE.x86_64 -> "x86_64"
-                MACOS_ARCHITECTURE.ARM64 -> "arm64-v8a"
+                CPU_ARCHITECTURE.x86_64 -> "x86_64"
+                CPU_ARCHITECTURE.ARM64 -> "arm64-v8a"
                 else -> throw IllegalStateException("Unsupported architecture $architecture")
             }.let {
                 "system-images;android-$version;google_apis;$it"
             }
             val abi = when (architecture) {
-                MACOS_ARCHITECTURE.x86_64 -> "x86_64"
-                MACOS_ARCHITECTURE.ARM64 -> "arm64-v8a"
+                CPU_ARCHITECTURE.x86_64 -> "x86_64"
+                CPU_ARCHITECTURE.ARM64 -> "arm64-v8a"
                 else -> throw IllegalStateException("Unsupported architecture $architecture")
             }
 

--- a/maestro-cli/src/main/java/maestro/cli/util/EnvUtils.kt
+++ b/maestro-cli/src/main/java/maestro/cli/util/EnvUtils.kt
@@ -122,7 +122,9 @@ object EnvUtils {
     }
 
     private fun determineArchitectureDetectionStrategy(): ArchitectureDetectionStrategy {
-        return if (runProcess("uname").contains("Linux")) {
+        return if (isWindows()) {
+            ArchitectureDetectionStrategy.WindowsArchitectureDetection
+        } else if (runProcess("uname").contains("Linux")) {
             ArchitectureDetectionStrategy.LinuxArchitectureDetection
         } else {
             ArchitectureDetectionStrategy.MacOsArchitectureDetection
@@ -172,6 +174,12 @@ sealed interface ArchitectureDetectionStrategy {
             }
         }
 
+    }
+
+    object WindowsArchitectureDetection: ArchitectureDetectionStrategy {
+        override fun detectArchitecture(): MACOS_ARCHITECTURE {
+            return MACOS_ARCHITECTURE.x86_64
+        }
     }
 }
 

--- a/maestro-cli/src/main/java/maestro/cli/util/EnvUtils.kt
+++ b/maestro-cli/src/main/java/maestro/cli/util/EnvUtils.kt
@@ -117,7 +117,7 @@ object EnvUtils {
         return Pair(first = version.getOrNull(), second = channel.getOrNull())
     }
 
-    fun getMacOSArchitecture(): MACOS_ARCHITECTURE {
+    fun getMacOSArchitecture(): CPU_ARCHITECTURE {
         return determineArchitectureDetectionStrategy().detectArchitecture()
     }
 
@@ -148,42 +148,42 @@ object EnvUtils {
 
 sealed interface ArchitectureDetectionStrategy {
 
-    fun detectArchitecture(): MACOS_ARCHITECTURE
+    fun detectArchitecture(): CPU_ARCHITECTURE
 
     object MacOsArchitectureDetection : ArchitectureDetectionStrategy {
-        override fun detectArchitecture(): MACOS_ARCHITECTURE {
+        override fun detectArchitecture(): CPU_ARCHITECTURE {
             fun runSysctl(property: String) = runProcess("sysctl", property).any { it.endsWith(": 1") }
 
             // Prefer sysctl over 'uname -m' due to Rosetta making it unreliable
             val isArm64 = runSysctl("hw.optional.arm64")
             val isX86_64 = runSysctl("hw.optional.x86_64")
             return when {
-                isArm64 -> MACOS_ARCHITECTURE.ARM64
-                isX86_64 -> MACOS_ARCHITECTURE.x86_64
-                else -> MACOS_ARCHITECTURE.UNKNOWN
+                isArm64 -> CPU_ARCHITECTURE.ARM64
+                isX86_64 -> CPU_ARCHITECTURE.x86_64
+                else -> CPU_ARCHITECTURE.UNKNOWN
             }
         }
     }
 
     object LinuxArchitectureDetection : ArchitectureDetectionStrategy {
-        override fun detectArchitecture(): MACOS_ARCHITECTURE {
+        override fun detectArchitecture(): CPU_ARCHITECTURE {
             return when (runProcess("uname", "-m").first()) {
-                "x86_64" -> MACOS_ARCHITECTURE.x86_64
-                "arm64" -> MACOS_ARCHITECTURE.ARM64
-                else -> MACOS_ARCHITECTURE.UNKNOWN
+                "x86_64" -> CPU_ARCHITECTURE.x86_64
+                "arm64" -> CPU_ARCHITECTURE.ARM64
+                else -> CPU_ARCHITECTURE.UNKNOWN
             }
         }
 
     }
 
     object WindowsArchitectureDetection: ArchitectureDetectionStrategy {
-        override fun detectArchitecture(): MACOS_ARCHITECTURE {
-            return MACOS_ARCHITECTURE.x86_64
+        override fun detectArchitecture(): CPU_ARCHITECTURE {
+            return CPU_ARCHITECTURE.x86_64
         }
     }
 }
 
-enum class MACOS_ARCHITECTURE {
+enum class CPU_ARCHITECTURE {
     x86_64,
     ARM64,
     UNKNOWN

--- a/maestro-cli/src/main/java/maestro/cli/util/WorkspaceUtils.kt
+++ b/maestro-cli/src/main/java/maestro/cli/util/WorkspaceUtils.kt
@@ -19,7 +19,7 @@ object WorkspaceUtils {
         val files = Files.walk(file).filter { !it.isDirectory() }.toList()
         val relativeTo = if (file.isDirectory()) file else file.parent
 
-        val outUri = URI.create("jar:file:${out.toAbsolutePath()}")
+        val outUri = URI.create("jar:${out.toUri()}")
         FileSystems.newFileSystem(outUri, mapOf("create" to "true")).use { fs ->
             files.forEach {
                 val outPath = fs.getPath(relativeTo.relativize(it).toString())

--- a/maestro-client/src/main/java/maestro/device/util/AndroidEnvUtils.kt
+++ b/maestro-client/src/main/java/maestro/device/util/AndroidEnvUtils.kt
@@ -92,12 +92,18 @@ object AndroidEnvUtils {
 
     private fun findCompatibleCommandLineTool(tool: String): File? {
         val path = File(androidHome, "cmdline-tools")
+
+        var thisTool = tool
+        if (EnvUtils.isWindows()){
+            thisTool = "$tool.bat"
+        }
+
         return path.listFiles()
-            ?.filter { it.isDirectory && File(it, "/bin/$tool").exists() }
-            ?.filter { isCommandLineToolCompatible(File(it, "bin/$tool")) }
+            ?.filter { it.isDirectory && File(it, "/bin/$thisTool").exists() }
+            ?.filter { isCommandLineToolCompatible(File(it, "bin/$thisTool")) }
             ?.sortedWith(compareBy<File> { it.name != "latest" }
                 .thenByDescending { it.name.toDoubleOrNull() })
-            ?.map { File(it, "bin/$tool") }
+            ?.map { File(it, "bin/$thisTool") }
             ?.firstOrNull()
     }
 

